errors/middleware-upgrade-guide.md

```diff
@@ -0,0 +1,310 @@
+
+The following patterns will no longer work:
+
+```js
+new Response('a text value')
+new Response(streamOrBuffer)
+new Response(JSON.stringify(obj), { headers: 'application/json' })
+```
+
+### How to upgrade
+
+For cases where Middleware is used to respond (such as authorization), you should migrate to use `rewrites`/`redirects` to pages that show an authorization error, login forms, or to an API Route.
+
+#### Before
+
+```typescript
+// pages/_middleware.ts
+import { NextResponse } from 'next/server'
+import type { NextRequest } from 'next/server'
+import { isAuthValid } from './lib/auth'
+
+export function middleware(request: NextRequest) {
+  // Example function to validate auth
+  if (isAuthValid(req)) {
+    return NextResponse.next()
+  }
+
+  return NextResponse.json({ message: 'Auth required' }, { status: 401 })
+}
+```
+
+#### After
+
+```typescript
+// middleware.ts
+import { NextResponse } from 'next/server'
+import type { NextRequest } from 'next/server'
+import { credentialsValid } from './lib/auth'
+
+export function middleware(request: NextRequest) {
+  const basicAuth = request.headers.get('authorization')
+
+  if (basicAuth) {
+    const auth = basicAuth.split(' ')[1]
+    const [user, pwd] = atob(auth).split(':')
+    // Example function to validate auth
+    if (credentialsValid(user, pwd)) {
+      return NextResponse.next()
+    }
+  }
+
+  const loginUrl = new URL('/login', request.url)
+  loginUrl.searchParams.set('from', request.nextUrl.pathname)
+
+  return NextResponse.redirect(loginUrl)
+}
+```
+
+## Cookies API Revamped
+
+### Summary of changes
+
+| Added                   | Removed       |
+| ----------------------- | ------------- |
+| `cookie.set`            | `cookie`      |
+| `cookie.delete`         | `clearCookie` |
+| `cookie.getWithOptions` | `cookies`     |
+
+### Explanation
+
+Based on beta feedback, we are changing the Cookies API in `NextRequest` and `NextResponse` to align more to a `get`/`set` model. The `Cookies` API extends Map, including methods like [entries](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map/entries) and [values](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map/entries).
+
+### How to upgrade
+
+`NextResponse` now has a `cookies` instance with:
+
+- `cookie.delete`
+- `cookie.set`
+- `cookie.getWithOptions`
+
+As well as other extended methods from `Map`.
+
+#### Before
+
+```javascript
+import { NextResponse } from 'next/server'
+import type { NextRequest } from 'next/server'
+
+export function middleware(request: NextRequest) {
+  // create an instance of the class to access the public methods. This uses `next()`,
+  // you could use `redirect()` or `rewrite()` as well
+  let response = NextResponse.next()
+  // get the cookies from the request
+  let cookieFromRequest = request.cookies['my-cookie']
```

Comment:
```suggestion
  let cookieFromRequest = request.cookies.get('my-cookie')
```

Reply:
From a discussion with @Kikobeats: prior to the change, request and response cookies were both plain objects.
This snipper is in the "Before" paragraph, so it should **not** use `.get()` nor `.set()`